local-cli/server/util/attachHMRServer.js

```diff
@@ -124,216 +127,43 @@ function attachHMRServer({httpServer, path, packagerServer}) {
-                  return;
-                }
-
-                const httpServerAddress = httpServer.address();
-
-                // Sanitize the value from the HTTP server
-                let packagerHost = 'localhost';
-                if (httpServer.address().address &&
-                    httpServer.address().address !== '::' &&
-                    httpServer.address().address !== '') {
-                  packagerHost = httpServerAddress.address;
-                }
-
-                return packagerServer.buildBundleForHMR({
-                  entryFile: client.bundleEntry,
-                  platform: client.platform,
-                  resolutionResponse,
-                }, packagerHost, httpServerAddress.port);
-              })
-              .then(bundle => {
-                if (!client || !bundle || bundle.isEmpty()) {
-                  return;
-                }
-
-                return JSON.stringify({
-                  type: 'update',
-                  body: {
-                    modules: bundle.getModulesIdsAndCode(),
-                    inverseDependencies: client.inverseDependenciesCache,
-                    sourceURLs: bundle.getSourceURLs(),
-                    sourceMappingURLs: bundle.getSourceMappingURLs(),
-                  },
-                });
-              })
-              .catch(error => {
-                // send errors to the client instead of killing packager server
-                let body;
-                if (error.type === 'TransformError' ||
-                    error.type === 'NotFoundError' ||
-                    error.type === 'UnableToResolveError') {
-                  body = {
-                    type: error.type,
-                    description: error.description,
-                    filename: error.filename,
-                    lineNumber: error.lineNumber,
-                  };
-                } else {
-                  console.error(error.stack || error);
-                  body = {
-                    type: 'InternalError',
-                    description: 'react-packager has encountered an internal error, ' +
-                      'please check your terminal error output for more details',
-                  };
-                }
-
-                return JSON.stringify({type: 'error', body});
-              })
-              .then(update => {
-                if (!client || !update) {
-                  return;
-                }
-
-                console.log(
-                  '[Hot Module Replacement] Sending HMR update to client (' +
-                  time() + ')'
-                );
-                client.ws.send(update);
-              });
-            },
-            () => {
-              // do nothing, file was removed
-            },
-          ).then(() => {
-            client.ws.send(JSON.stringify({type: 'update-done'}));
+          onDisconnect,
+        }));
+        if (clients.length === 1) {
```

Comment:
The idea is that we only add a listener when the first client is registered, since this indicates to the packager that we want to enter "hot mode", which means we can't register the listener before any clients request it.
I think I can add a flag instead, if you think this isn't explicit.